benchmarks/polys.py

```diff
@@ -1,17 +1,132 @@
-import sympy
+from sympy import symbols, prod, prem
+from sympy.polys import ZZ, Poly
+
 
 class TimePolyManyGens:
     """Time using a Poly with many generators"""
 
     params = [1, 10, 100, 500]
 
     def setup(self, n):
-        self.xs = sympy.symbols('x:{}'.format(n))
+        self.xs = symbols('x:{}'.format(n))
         self.x = self.xs[n // 2]
-        self.px = sympy.Poly(self.x, self.xs)
+        self.px = Poly(self.x, self.xs)
 
     def time_create_poly(self, n):
-        sympy.Poly(self.x, self.xs)
+        Poly(self.x, self.xs)
 
     def time_is_linear(self, n):
         self.px.is_linear
+
+
+class PolyPrem:
+    """Benchmark for the prem method"""
+
+    def setup(self, n):
+        self.x, *self.y = symbols("x, y1:11")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+
+        self.values = {}
+
+        self.f_case1, self.g_case1 = self.generate_poly_case1(n)
+        self.f_case2, self.g_case2 = self.generate_poly_case2(n)
+        self.f_case3, self.g_case3 = self.generate_poly_case3(n)
+        self.f_case4, self.g_case4 = self.generate_poly_case4(n)
+
+        self.fp_case1 = Poly(self.f_case1, self.x, *(self.y)[:n])
+        self.gp_case1 = Poly(self.g_case1, self.x, *(self.y)[:n])
+        self.fpe_case1 = self.R.from_sympy(self.f_case1)
+        self.gpe_case1 = self.R.from_sympy(self.g_case1)
+
+        self.fp_case2 = Poly(self.f_case2, self.x, *(self.y)[:n])
+        self.gp_case2 = Poly(self.g_case2, self.x, *(self.y)[:n])
+        self.fpe_case2 = self.R.from_sympy(self.f_case2)
+        self.gpe_case2 = self.R.from_sympy(self.g_case2)
+
+        self.fp_case3 = Poly(self.f_case3, self.x, *(self.y)[:n])
+        self.gp_case3 = Poly(self.g_case3, self.x, *(self.y)[:n])
+        self.fpe_case3 = self.R.from_sympy(self.f_case3)
+        self.gpe_case3 = self.R.from_sympy(self.g_case3)
+
+        self.fp_case4 = Poly(self.f_case4, self.x, *(self.y)[:n])
+        self.gp_case4 = Poly(self.g_case4, self.x, *(self.y)[:n])
+        self.fpe_case4 = self.R.from_sympy(self.f_case4)
+        self.gpe_case4 = self.R.from_sympy(self.g_case4)
+
+    def generate_poly_case1(self, n):
+        """Linearly dense quartic inputs with quadratic GCDs."""
+        D = (1 + self.x + sum((self.y)[:n])) ** 2
+        f = D * (-2 + self.x - sum((self.y)[:n])) ** 2
+        g = D * (2 + self.x + sum((self.y)[:n])) ** 2
+        return f, g
+
+    def generate_poly_case2(self, n):
+        """Sparse GCD and inputs where degree is proportional to the number of variables."""
+        D = 1 + self.x ** (n + 1) + sum([(self.y)[i] ** (n + 1) for i in range(n)])
+        f = D * (-2 + self.x ** (n + 1) + sum([(self.y)[i] ** (n + 1) for i in range(n)]))
+        g = D * (2 + self.x ** (n + 1) + sum([(self.y)[i] ** (n + 1) for i in range(n)]))
+        return f, g
```

Comment:
I think that it would be better to make separate classes for each of these cases like:
```python
class LinearDenseQuadraticGCD:
    def generate(self, n):
        ...

class SparseGCDHighDegree:
    def generate(self, n):
        ...
```
Then you could avoid the repetitive code that lists all the cases and the timing output would be a bit more meaningful than just "case1", "case2" etc.:
```
[ 41.24%] ··· ...TimePolyPremFast.time_PolyElement_prem_case1                 ok
[ 41.24%] ··· ======== =============
               param1               
              -------- -------------
                 1      11.7±0.08ms 
                 5       94.5±0.9ms 
                 8        178±2ms   
                 10       186±1ms   
              ======== =============

[ 41.40%] ··· ...TimePolyPremFast.time_PolyElement_prem_case2                 ok
[ 41.40%] ··· ======== ============
               param1              
              -------- ------------
                 1      5.73±0.2ms 
                 5      36.2±0.7ms 
                 8       83.6±2ms  
                 10      101±2ms   
              ======== ============
```

Reply:
Should I include the `time_` function for the respective cases in the same class or in different
like this ?

> ```python
> class LinearDenseQuadraticGCD:
>     def generate(self, n):
>
>     def time_(self, n):

Reply:
There are different ways to do it. The most important thing is how useful the asv output is. The things that we would want are to see how different methods perform relative to each other and how each method performs as n scales up.

Another way is to have pairs of parameters like
```
params = [(1, 1), (1, 2), (1, 5), (2, 1), (2, 2), ...]
```